overlord/snapstate/autorefresh.go

```diff
@@ -873,11 +887,90 @@ func maybeAddRefreshInhibitNotice(st *state.State) error {
 		}
 	}
 
-	if changed {
-		if _, err := st.AddNotice(nil, state.RefreshInhibitNotice, "-", nil); err != nil {
+	if err := maybeAddRefreshInhibitWarningFallback(st, curInhibitedSnaps, changed); err != nil {
+		return err
+	}
+
+	if !changed {
+		// the set of inhibited snaps was not changed, nothing to do.
+		return nil
+	}
+
+	if _, err := st.AddNotice(nil, state.RefreshInhibitNotice, "-", nil); err != nil {
+		return err
+	}
+	st.Set("last-recorded-inhibited-snaps", curInhibitedSnaps)
+
+	return nil
+}
+
+// maybeAddRefreshInhibitWarningFallback records a warning if the set of
+// inhibited snaps was changed since the last notice.
+//
+// The warning is recorded only if:
+// 	1. There is at least 1 inhibited snap.
+// 	2. The "refresh-app-awareness-ux" experimental flag is enabled.
+// 	3. No snap exists with the marker "snap-refresh-observe" interface connected.
+func maybeAddRefreshInhibitWarningFallback(st *state.State, inhibitedSnaps map[string]bool, changed bool) error {
+	if len(inhibitedSnaps) == 0 {
+		// do nothing
+		return nil
+	}
+
+	tr := config.NewTransaction(st)
+	experimentalRefreshAppAwarenessUX, err := features.Flag(tr, features.RefreshAppAwarenessUX)
+	if err != nil && !config.IsNoOption(err) {
+		return err
+	}
+	if !experimentalRefreshAppAwarenessUX {
+		// snapd will send notifications directly, check maybeAsyncPendingRefreshNotification
+		return nil
+	}
+
+	markerExists, err := HasActiveConnection(st, "snap-refresh-observe")
+	if err != nil {
+		return err
+	}
+
+	if markerExists {
+		// do nothing
+		return nil
+	}
+
+	// let's fallback to issuing warnings if no snap exists with the
+	// marker snap-refresh-observe interface connected.
+
+	// building warning message
+	var snapsBuf bytes.Buffer
+	i := 0
+	for snap := range inhibitedSnaps {
+		if i > 0 {
+			snapsBuf.WriteString(", ")
+		}
+		snapsBuf.WriteString(snap)
+		i++
+	}
+	message := fmt.Sprintf("cannot refresh (%s) due running apps; close running apps to continue refresh.", snapsBuf.String())
+
+	// replace warning message if there is an existing inhibition warning.
+	for _, warning := range st.AllWarnings() {
+		if !strings.HasSuffix(warning.String(), "close running apps to continue refresh.") {
+			continue
+		}
+		// match found, replace
+		// XXX: is it worth it to check for unexpected multiple matches?
+		if err := st.ReplaceWarningMessage(warning.String(), message); err != nil {
 			return err
 		}
-		st.Set("last-recorded-inhibited-snaps", curInhibitedSnaps)
+		break
+	}
+
+	if changed {
+		// set of inhibited snaps was changed, issue a warning without cool down.
+		st.AddWarning(message, nil)
+	} else {
+		// wait some time before showing the same warning to the user again after okaying.
+		st.AddWarning(message, &state.AddWarningOptions{RepeatAfter: 6 * time.Hour})
```

Comment:
didn't we discuss waiting a full day?

Reply:
I must have misremembered. Thanks! Updated